### PR TITLE
renderPage API for top-level component rendering & layout feature implementation

### PR DESCRIPTION
**Pull request by tz5514**
This PR implement the `renderPage` approach in Next.js core, according to discussing in https://github.com/zeit/next.js/pull/3461#issuecomment-352631686. & https://github.com/zeit/next.js/pull/3471

This way can make core API of Next.js stay bare minimal, yet flexible.
With this core API, we can define a static property "renderPage" of page component.
It will be used to decide how page render at top level.

ex:
```js
class IndexPage extends React.Component {
  static renderPage({ Component, props, url }) {
    return (
      <AppLayout>
        <Component {...props} url={url} />
      </AppLayout>
    )
  }

  render () {
    return (
      <p>Index</p>
    )
  }
}
```
And this is enough for us to implement any extra features for persistent layout outside of Next.js core.

So I also implement a sample of third-party extension for layout feature [in this PR](https://github.com/zeit/next.js/pull/3552/files#diff-2f826f233c68478d2b3457c14c12be0b) (outside of Next.js core).

It support the multiple layer, layout's getInitialProps, and getInitialProps's concurrent mode.
so we can do something like:
```js
// layouts/AppLayout.js
import React from 'react'
import delay from '../modules/delay'

export default class AppLayout extends React.Component {
  static async getInitialProps () {
    return {
      delay: await delay(1000)
    }
  }

  render () {
    return (
      <div>
        <h1>App header</h1>
        {this.props.children}
      </div>
    )
  }
}
```

```js
// layouts/ContentLayout.js
import AppLayout from './AppLayout'

export default class ContentLayout extends React.Component {
  static parentLayout = AppLayout
  static async getInitialProps () {
    return {
      delay: await delay(2000)
    }
  }

  render () {
    return (
      <div>
        <hr />
        {this.props.children}
        <hr />
      </div>
    )
  }
}

```

```js
// pages/about.js
import React from 'react'
import delay from '../modules/delay'
import applyLayout from '../modules/next-layouts'
import ContentLayout from '../layouts/ContentLayout'

class AboutPage extends React.Component {
  static layout = ContentLayout
  static async getInitialProps () {
    return {
      delay: await delay(3000)
    }
  }

  render () {
    return (
      <p>about</p>
    )
  }
}

export default applyLayout(AboutPage, {
  getInitialPropsMode: 'concurrent'
})
```

and the output page view will be:
image

If this approach will be accepted, I think we can make this module to be a npm package that naming "next-layouts".

### CODE REVIEW

**Review by frol — APPROVED** [review]
This should close #88! LGTM!

@tz5514 Thank you!

**Comment by sergiodxa** [comment]
Why `static layout` and `static parentLayout` instead of just `static layout` and compose them N levels? So page X can have LayoutA and LayoutA can have a LayoutB applied and they can just be composed.

**Comment by frol** [comment]
@sergiodxa The example demonstrates that given this small change in the core, it is possible to implement quite an advanced third-party API.

Should this advanced example be extracted into a separate PR?

**Comment by tz5514** [comment]
@sergiodxa @frol Yeah This layout feature is just a draft demo that can be discussed independently. Mainly idea in this PR is `renderPage` API of Next.js core. 
If you guys feel confusing, I think I can move the example to a separate PR 😃 .

**Comment by frol** [comment]
@timneutkens Could you review/merge this?

**Comment by timneutkens** [comment]
We'll discuss/review this PR after v5 is released. Since it's a rather big semantic change and we've been discussing it internally too. Thanks for your patience 🙏

**Comment by Nishchit14** [comment]
@timneutkens Thanks for looking into this issue, and When we're releasing v5?

**Comment by frol** [comment]
> Since it's a rather big semantic change

I just want to point out that this PR does not *change* the existing behavior, it *adds* new API. While I personally believe that this new API should be used in 99% of the Next.js projects, it is not a breaking change and it is still a valid React JS code, so if you decide on a better default API (which will be a breaking change), it shouldn't be an issue to migrate from this `renderPage` API to a new one.

**Comment by imagine10255** [comment]
Will the v5 beta already contains the renderPage API?

**Comment by juciccio** [comment]
Guys I have a problem when I try to install this PR. The npm install fails when it tries to build the next.js module.

**Comment by RustyDev** [comment]
Has anyone tried this with solution with Redux? Running into an issue when trying to wrap the AppLayout using withRedux (`next-redux-wrapper`) so the provider is top-level. `getInitialProps` is getting called before the redux store in initiated so SSR isn't working quite right. Just want to make sure that this has been tested in that scenario.

```
// AppLayout

export default withRedux(configureStore, mapStateToProps, mapDispatchToProps)(
  AppLayout
);

// reducer

export function fetchPlaylist(playlist, url) {
  return dispatch => {
    return fetch(url)
      .then(response => response.json())
      .then(
        response => {
          const { result, entities } = normalize(response.tracks, [
            trackSchema,
          ]);
          dispatch(
            fetchPlaylistSuccess( playlist, response.title, response.meta, result, entities )
          );
        }
      );
  };
}

//

Popular.getInitialProps = async function({ store, isServer, pathname, query }) {

  // ...

  const initData = await store.dispatch(fetchPlaylist(playlist, url));

  return initData
}

```

**Comment by frol** [comment]
@juciccio git repo doesn't have pre-compiled `dist` folder, so you cannot install Next.js directly from the repo with `npm install <repo>`. Follow the guide from the [`contributing.md`](https://github.com/zeit/next.js/blob/canary/contributing.md).

@RustyDev Elaborate more on "SSR isn't working quite right." As far as I can see, you expect `store` to be passed to `getInitialProps`, but I am not sure if that has ever been the case.

**Comment by RustyDev** [comment]
@frol Using [next-redux-wrapper](https://github.com/kirill-konshin/next-redux-wrapper), it's possible to have the store passed into `getInitialProps` via it's HOC. When trying to use SSR and Redux using AppLayout, I'm able to console.log and see the dispatch happen, but it looks like the store is possibly being re-instantiated afterwards. Maybe this is out of the scope of this discussion, but I was hoping to have a solution with Redux where I can wrap the app using AppLayout and then have the store be accessible in the top-level layout. Basically like this so that the page, header and footer all have access. The nice thing about this implementation is that the footer holds an audio player and is persistent between pages.

```html
<App>
  <Provider>
    <AppLayout>
      <HeaderContainer/>
      {...this.props.children}
      <FooterContainer/>
    </AppLayout>
  </Provider>
</App>
```
I'll put together a simple example to show what's happening. I feel like I won't be the only one who will want this ability and see this as the solution.

**Comment by RustyDev** [comment]
Here's an [example repo](https://github.com/RustyDev/next-with-redux-applayout)

It's deployed [here](https://with-redux-yfrswhwlpz.now.sh) where you can see the store reset to 00:00:00

This is the intended [result](https://with-redux-dwzgszejiy.now.sh/)

**Comment by frol** [comment]
@RustyDev https://github.com/RustyDev/next-with-redux-applayout/pull/1 - in this PR I made it working and this proves that `renderPage` API plays just fine with Redux (and, probably, the rest of the React ecosystem). It also shows that the magic of `next-layout` and `next-redux-wrapper` needs to be extended to handle `renderPage` API correctly. To be honest, I am not a fan of the magical solutions so the plain `renderPage` with [explicit Redux integration as it was in the demos before `next-redux-wrapper`](https://github.com/zeit/next.js/tree/4a73ccbb00d6fa112bc8e0763dfceeade263033f/examples/with-redux) work much better for me.

**Comment by frol** [comment]
@timneutkens @arunoda ping

**Comment by rashidul0405** [comment]
@frol RustyDev/next-with-redux-applayout#1 doesn't look like a good solution, *global* store should be connected to top-level component instead of every page.

**Comment by frol** [comment]
@rashidul0405 Is "should be" related to correctness, performance or what? `next-redux-wrapper` is designed to wrap pages (it stores the global store in its own "global" variable, which is then attached to pages, so it doesn't re-create the store). You may want to play with `_document.js` API if you want to have something truly global.

**Comment by frol** [comment]
@timneutkens

> We'll discuss/review this PR after v5 is released.

(c) https://github.com/zeit/next.js/pull/3552#issuecomment-360577112

What are the particular issues with this PR? Do you have a better idea on how to tackle #88? Can you unlock people who need this issue to be resolved?

/cc @arunoda

**Comment by linasuk** [comment]
Hello, v5 has been released is there any updates on this?
my current solution is far from pretty and I'd love to hear some better solutions.

**Comment by krazyjakee** [comment]
What does this addition give that this example does not? https://github.com/zeit/next.js/tree/canary/examples/layout-component

Am I able to maintain the state of a layout over multiple routes without re-mounting it?

**Comment by isBatak** [comment]
@krazyjakee if you rewrite layout component like this https://gist.github.com/isBatak/9ee6ca051ed3ea3435af31b2c7c40d2f you will see that it is actually re-mounting on each page change.

**Comment by davixyz** [comment]
Any updates on when is this going to be merged? currently doing some work-arounds to achieve this preservation of state of a single component when navigating in multiple routes.

**Comment by Nishchit14** [comment]
**It's time to move on now**
This issue is the heart of the Nextjs, but still, people are focusing on v5 rather than solving the real problem.

Next is too good, we can develop small to medium size of the application, but for the enterprise level application, It is still risky. 

Thank you, Guys.

**Comment by tz5514** [comment]
@Nishchit14 I can't agree with you more. Honestly, I'm also disappointed that they keep silent on this issue. I think the solution in this PR is stand by lots of people, but they still
adhere to discussing this feature in internal, and didn't give it a priority.

Next.js is awesome, and one the most important project in React ecosystem. We really thanks for core team's work. But this issue really stagnate too long.

**Comment by timneutkens** [comment]
It's on my list for Next v6. Sorry for taking longer, like I said before:

> Since it's a rather big semantic change and we've been discussing it internally too. Thanks for your patience 🙏

To give some context:

I'm basically focusing on making 5.x very stable before moving on to introducing big new things (like this one) 👍 5.1 is coming very soon 👍, and after that I can focus on multiple things, including this. I'm very aware of how many people want this feature.

> It's time to move on now
This issue is the heart of the Nextjs, but still, people are focusing on v5 rather than solving the real problem.

It's a twofold, if we gave you this feature instead of working on improvement of v5 you, or someone else, would go ahead and say: my hot reloading state is broken. Or whatever other issues there were.

**Comment by Nishchit14** [comment]
@timneutkens thank you for the breaking silence, **feeling better now 😃**

@tz5514 I agree with you sir, I respect the core team and very thankful to them, But due to long silence on this issue I was feeling like a frozen in Iceland. It's never been easier to migrate to the new framework and when you stuck in between then It is more painful. 

Thank you team again :)

**Comment by glemiere** [comment]
I love this PR. It's a total must have!
@tz5514 is there a way we can start working on your idea of a next-layouts package until it's merged ?
@timneutkens when do you plan to release the version 6 ?

**Comment by timneutkens** [comment]
Closing in favor of #4129

**Pull request closed without merging**
